lib/model/model.go

```diff
@@ -2685,6 +2698,66 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 	return true
 }
 
+func (m *model) cleanPending(cfg config.Configuration, removedFolders map[string]bool) {
+	// Cache to maps for easy lookups
+	ignoredDevices := make(map[protocol.DeviceID]bool, len(cfg.IgnoredDevices))
+	for _, dev := range cfg.IgnoredDevices {
+		ignoredDevices[dev.ID] = true
+	}
+	existingDevices := cfg.DeviceMap()
+	existingFolders := make(map[string]*config.FolderConfiguration, len(cfg.Folders))
+	for i := range cfg.Folders {
+		folder := &cfg.Folders[i]
+		existingFolders[folder.ID] = folder
+	}
+
+	pendingFolders, err := m.db.PendingFolders(protocol.EmptyDeviceID)
+	if err != nil {
+		l.Infof("Could not iterate through pending folder entries for cleanup: %v", err)
+	}
+	for folderID, offers := range pendingFolders {
+		if _, ok := removedFolders[folderID]; ok {
+			// Forget pending folder device associations for recently removed
+			// folders as well, assuming the folder is no longer of interest
+			// at all (but might become pending again).
+			l.Debugf("Discarding pending removed folder %v from all devices", folderID)
+			m.db.RemovePendingFolder(folderID, protocol.EmptyDeviceID)
+			continue
+		}
+		for deviceID := range offers {
+			if dev, ok := existingDevices[deviceID]; !ok {
+				l.Debugf("Discarding pending folder %v from unknown device %v", folderID, deviceID)
+				m.db.RemovePendingFolder(folderID, deviceID)
+				continue
+			} else if dev.IgnoredFolder(folderID) {
+				l.Debugf("Discarding now ignored pending folder %v for device %v", folderID, deviceID)
+				m.db.RemovePendingFolder(folderID, deviceID)
+				continue
+			}
+			if folderCfg, ok := existingFolders[folderID]; ok {
+				if folderCfg.SharedWith(deviceID) {
+					l.Debugf("Discarding now shared pending folder %v for device %v", folderID, deviceID)
+					m.db.RemovePendingFolder(folderID, deviceID)
+				}
+			}
+		}
+	}
+
+	pendingDevices, err := m.db.PendingDevices()
```

Comment:
Error not looked at, some behavior probably missing?

Reply:
Right, thanks for catching that.